src/Twilio/Security/RequestValidator.cs

```diff
@@ -124,23 +127,48 @@ private static bool SecureCompare(string a, string b)
             return mismatch == 0;
         }
 
-        private string RemovePort(UriBuilder uri)
+        private string RemovePort(string url)
         {
-            // UriBuilder.ToString() will not display the port 
-            // if the Port property is set to -1
+            var uri = new UriBuilder(url);
+            uri.Host = PreserveHostnameCasing(url);
             uri.Port = -1;
-            return uri.ToString();
+            var scheme = PreserveSchemeCasing(url);
+            return uri.Uri.OriginalString.Replace(uri.Scheme, scheme);
         }
 
-        private string AddPort(UriBuilder uri)
+        private string AddPort(string url)
         {
+            var uri = new UriBuilder(url);
+            uri.Host = PreserveHostnameCasing(url);
             if (uri.Port != -1)
             {
-                return uri.ToString();
+                return uri.Uri.OriginalString;
             }
+            var scheme = PreserveSchemeCasing(url);
             uri.Port = uri.Scheme == "https" ? 443 : 80;
-            return uri.ToString();
+            return uri.Uri.OriginalString.Replace(uri.Scheme, scheme);
         }
 
+        private string PreserveHostnameCasing(string url)
+        {
+            var m = Regex.Match(url, URL_REGEX);
+            var hostName = m.Groups[4].ToString();
+            var preservedHostName = System.String.Empty;
+            // Do we have credentials in the URL?
+            if (hostName.Split('@').Length == 2)
+            {
+                preservedHostName = hostName.Split('@')[1].Split(':')[0];
+            } else {
+                preservedHostName = hostName.Split(':')[0];
+            }
+            return preservedHostName;
```

Comment:
Think rather than trying to regex parse and extract out pieces of the URL (which is what `UriBuilder` is also doing), the logic can be simplified by getting the substring from the original URL case insensitive. E.g., `uri.Host = url.substring(url.toLowerCase().indexOf(uri.Host), uri.Host.length());` Just in C# and not Java syntax.